### PR TITLE
Update CSPEC to use the CAEN readout for 3He detectors

### PR DESCRIPTION
**Pull request by g5t**
### #623 623
The detector for CSPEC will be comprised of <sup>3</sup>He MultiTubes instead of MultiGrids, move the implementation to use the common CAEN machinery.

## Checklist for submitter

- [ ] Check for conflict with integration test
- [ ] Unit tests pass

---

### Nominate for Group Code Review

- [ ] Nominate for code review

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -75,8 +77,9 @@ add_custom_target(runefu
     COMMAND perfgen "-s" "1" "--nohwcheck"
     COMMAND dream "-f" "../configs/DreamEndcap.json" "-s" "2" "--nohwcheck"
     COMMAND freia "-f" "../configs/freia.json" "-s" "2" "--nohwcheck" "-l" "Info" "--log_file" "deleteme_graylog.log"
-    COMMAND cspec "-f" "../configs/cspec.json" "-s" "2" "--nohwcheck"
+    COMMAND trex "-f" "../configs/tex.json" "-s" "2" "--nohwcheck"
     COMMAND ttlmonitor "-f" "../configs/freiamon.json" "-s" "2" "--nohwcheck" "--dumptofile" "deleteme_ttlmon_" "-t" "freia_beam_monitor"
     COMMAND nmx "-f" "../configs/nmx.json" "-s" "2"  "--nohwcheck"
-    DEPENDS copycfgs dream loki bifrost miracles freia cspec ttlmonitor nmx
+    COMMAND cspec "-f" "../configs/he3cpec.json" "-s" "2" "--nohwcheck" "--dumptofile" "deleteme_cspec.h5"
```
Comment (mortenjc) [inline src/CMakeLists.txt:83]: I suggest keeping the old cspec.json name

File: src/modules/cspec/CMakeLists.txt [hunk 2]
```diff
@@ -1,58 +1,12 @@
-# Copyright (C) 2021 European Spallation Source, ERIC. See LICENSE file
-
-add_subdirectory(test)
-add_subdirectory(generators)
-include_directories(.)
+# Copyright (C) 2022 European Spallation Source, ERIC. See LICENSE file
```
Comment (mortenjc) [inline src/modules/cspec/CMakeLists.txt:1]: 2022 - 2023 ?
If this started in 2022, though, else just 2023

File: src/modules/cspec/geometry/CspecGeometry.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+// #define TRC_LEVEL TRC_L_DEB
+
+namespace Caen {
+class CspecGeometry : public Geometry {
+public:
+    CspecGeometry(Config &CaenConfiguration);
+  uint32_t calcPixel(DataParser::CaenReadout &Data);
+  bool validateData(DataParser::CaenReadout &Data);
+
+  int PackTubes{24}; // tubes per triplet (might be obvious from the name)
```
Comment (mortenjc) [inline src/modules/cspec/geometry/CspecGeometry.h:28]: comment seems off?

File: src/modules/cspec/geometry/CspecGeometry.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  /// \brief return local x-coordinate from amplitudes
+  int xCoord(int AmpA, int AmpB) {
+    int Coord = reverse(posAlongTube(AmpA, AmpB)) % (NPos / 3);
+    XTRACE(DATA, DEB, "AmpA %d, AmpB %d, xCoord %d", AmpA, AmpB, Coord);
+    return Coord;
+  }
+
+  /// \brief return local y-coordinate from amplitudes
+  int yCoord(int AmpA, int AmpB) {
+    return (reverse(posAlongTube(AmpA, AmpB)) * PackTubes) / NPos;
```
Comment (mortenjc) [inline src/modules/cspec/geometry/CspecGeometry.h:45]: there seem to be some assumptions going on here that I don't know about.

But given the height of CSPEC I would expect only a single tube per (A,B) pair?
If this is the case then the position calculations are wrong. If the assumption is wrong and 
you have inside information we need to update the relevant ICD's accoringly.

File: src/modules/cspec/geometry/CspecGeometry.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+  /// \brief return the position along the tube
+  int posAlongTube(int AmpA, int AmpB);
+
+  /// NullCalibration just reverses the middle position
+  int reverse(int OrgPos) {
+    if ((OrgPos < 100) or (OrgPos >= 200)) {
+      return OrgPos;
+    }
+    return 299 - OrgPos;
```
Comment (mortenjc) [inline src/modules/cspec/geometry/CspecGeometry.h:56]: also the 300 pixel resolution for a (bifrost) triplet is likely to be wrong here.

File: src/modules/trex/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+add_subdirectory(generators)
+include_directories(.)
+
+#=============================================================================
+# trex detector module and base test
+#=============================================================================
+set(trex_common_inc
+        TREXBase.h
+        TREXInstrument.h
+  geometry/Config.h
```
Comment (mortenjc) [inline src/modules/trex/CMakeLists.txt:13]: inconsistent indentation

**Review by mortenjc — APPROVED** [review]
I have no big issues, mostly this is just search and replace from CSPEC to TREX.

There are a few small suggestions.

File: src/CMakeLists.txt (hunk 1 above)
Comment (g5t) [reply src/CMakeLists.txt:83]: This is an oversight from the string replacement. I'll change it.

File: src/modules/cspec/CMakeLists.txt (hunk 2 above)
Comment (g5t) [reply src/modules/cspec/CMakeLists.txt:1]: Another casualty from excessive `sed` usage.

File: src/modules/cspec/geometry/CspecGeometry.h (hunk 4 above)
Comment (g5t) [reply src/modules/cspec/geometry/CspecGeometry.h]: The assumptions are all BIFROST based (I copied that directory as a base and didn't modify the geometry information sufficiently). The latest commit 'fixes' this by punting on most of it, and assuming that the detector will be cylindrical with all tubes at constant radius and height and all the same length.

Yes, it's very likely that every tube will give a single (A,B) pair.

File: src/modules/cspec/geometry/CspecGeometry.h (hunk 5 above)
Comment (g5t) [reply src/modules/cspec/geometry/CspecGeometry.h]: Flipping the centre of the tube is certainly wrong, whoops! 
300 pixels is probably wrong as well if the detector is 3.5 metres long; but this is not certain as far as I'm aware.

**Pull request merged**